compiler/src/dotty/tools/dotc/transform/ShortcutImplicits.scala

```diff
@@ -0,0 +1,165 @@
+     */
+    private def directMethod(sym: Symbol)(implicit ctx: Context): Symbol =
+      if (sym.owner.isClass) {
+        val direct = sym.owner.info.member(sym.name.directName)
+          .suchThat(_.info matches directInfo(sym.info)).symbol
+        if (direct.maybeOwner == sym.owner) direct
+        else newDirectMethod(sym).enteredAfter(thisTransform)
+      }
+      else directMeth.getOrElseUpdate(sym, newDirectMethod(sym))
+
+
+    /** Transform `qual.apply` occurrences according to rewrite rule (2) above */
+    override def transformSelect(tree: Select)(implicit ctx: Context, info: TransformerInfo) =
+      if (tree.name == nme.apply &&
+          defn.isImplicitFunctionType(tree.qualifier.tpe.widen) &&
+          shouldBeSpecialized(tree.qualifier.symbol)) {
+        def directQual(tree: Tree): Tree = tree match {
+          case Apply(fn, args)     => cpy.Apply(tree)(directQual(fn), args)
+          case TypeApply(fn, args) => cpy.TypeApply(tree)(directQual(fn), args)
+          case Block(stats, expr)  => cpy.Block(tree)(stats, directQual(expr))
+          case tree: RefTree =>
+            cpy.Ref(tree)(tree.name.directName)
+              .withType(directMethod(tree.symbol).termRef)
+        }
+        directQual(tree.qualifier)
+      } else tree
+
+    /** Transform methods with implicit function type result according to rewrite rule (1) above */
+    override def transformDefDef(mdef: DefDef)(implicit ctx: Context, info: TransformerInfo): Tree = {
+      val original = mdef.symbol
+      if (shouldBeSpecialized(original)) {
+        val direct = directMethod(original)
+
+        def splitClosure(tree: Tree): (List[Type] => List[List[Tree]] => Tree, Tree) = tree match {
+          case Block(Nil, expr) => splitClosure(expr)
+          case Block((meth @ DefDef(nme.ANON_FUN, Nil, clparams :: Nil, _, _)) :: Nil, cl: Closure) =>
+            val tparamSyms = mdef.tparams.map(_.symbol)
+            val vparamSymss = mdef.vparamss.map(_.map(_.symbol))
+            val clparamSyms = clparams.map(_.symbol)
+            val remappedCore = (ts: List[Type]) => (prefss: List[List[Tree]]) =>
+              meth.rhs
+                .subst(tparamSyms ::: (vparamSymss.flatten ++ clparamSyms),
+                       ts.map(_.typeSymbol) ::: prefss.flatten.map(_.symbol))
+                .changeOwnerAfter(original, direct, thisTransform)
+                .changeOwnerAfter(meth.symbol, direct, thisTransform)
+            val forwarder = ref(direct)
+              .appliedToTypeTrees(tparamSyms.map(ref(_)))
+              .appliedToArgss(vparamSymss.map(_.map(ref(_))) :+ clparamSyms.map(ref(_)))
+            val fwdClosure = cpy.Block(tree)(cpy.DefDef(meth)(rhs = forwarder) :: Nil, cl)
+            (remappedCore, fwdClosure)
+          case EmptyTree =>
+            (_ => _ => EmptyTree, EmptyTree)
+        }
+
+        val (remappedCore, fwdClosure) = splitClosure(mdef.rhs)
+        val originalDef = cpy.DefDef(mdef)(rhs = fwdClosure)
```

Comment:
Do you need to reset the `ABSTRACT` flag on the original method, now that it always contains a forwarder?